pkg/framework/plugins/removepodsviolatingnodeaffinity/node_affinity.go

```diff
@@ -0,0 +1,120 @@
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"context"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/descheduler/pkg/apis/componentconfig"
+	"sigs.k8s.io/descheduler/pkg/descheduler/evictions"
+	nodeutil "sigs.k8s.io/descheduler/pkg/descheduler/node"
+	podutil "sigs.k8s.io/descheduler/pkg/descheduler/pod"
+	"sigs.k8s.io/descheduler/pkg/framework"
+)
+
+const PluginName = "RemovePodsViolatingNodeAffinity"
+
+// RemovePodsViolatingNodeAffinity evicts pods on the node which violate node affinity
+type RemovePodsViolatingNodeAffinity struct {
+	handle    framework.Handle
+	args      *componentconfig.RemovePodsViolatingNodeAffinityArgs
+	podFilter podutil.FilterFunc
+}
+
+var _ framework.Plugin = &RemovePodsViolatingNodeAffinity{}
+var _ framework.DeschedulePlugin = &RemovePodsViolatingNodeAffinity{}
+
+// New builds plugin from its arguments while passing a handle
+func New(args runtime.Object, handle framework.Handle) (framework.Plugin, error) {
+	nodeAffinityArgs, ok := args.(*componentconfig.RemovePodsViolatingNodeAffinityArgs)
+	if !ok {
+		return nil, fmt.Errorf("want args to be of type RemovePodsViolatingNodeAffinityArgs, got %T", args)
+	}
+
+	var includedNamespaces, excludedNamespaces sets.String
+	if nodeAffinityArgs.Namespaces != nil {
+		includedNamespaces = sets.NewString(nodeAffinityArgs.Namespaces.Include...)
+		excludedNamespaces = sets.NewString(nodeAffinityArgs.Namespaces.Exclude...)
+	}
+
+	podFilter, err := podutil.NewOptions().
+		WithFilter(handle.Evictor().Filter).
+		WithNamespaces(includedNamespaces).
+		WithoutNamespaces(excludedNamespaces).
+		WithLabelSelector(nodeAffinityArgs.LabelSelector).
+		BuildFilterFunc()
+	if err != nil {
+		return nil, fmt.Errorf("error initializing pod filter function: %v", err)
+	}
+
+	return &RemovePodsViolatingNodeAffinity{
+		handle:    handle,
+		podFilter: podFilter,
+		args:      nodeAffinityArgs,
+	}, nil
+}
+
+// Name retrieves the plugin name
+func (d *RemovePodsViolatingNodeAffinity) Name() string {
+	return PluginName
+}
+
+func (d *RemovePodsViolatingNodeAffinity) Deschedule(ctx context.Context, nodes []*v1.Node) *framework.Status {
+	for _, nodeAffinity := range d.args.NodeAffinityType {
+		klog.V(2).InfoS("Executing for nodeAffinityType", "nodeAffinity", nodeAffinity)
+
+		switch nodeAffinity {
+		case "requiredDuringSchedulingIgnoredDuringExecution":
+		loop:
+			for _, node := range nodes {
+				klog.V(1).InfoS("Processing node", "node", klog.KObj(node))
+
+				pods, err := podutil.ListPodsOnANode(
+					node.Name,
+					d.handle.GetPodsAssignedToNodeFunc(),
+					podutil.WrapFilterFuncs(d.podFilter, func(pod *v1.Pod) bool {
+						return d.handle.Evictor().Filter(pod) &&
+							!nodeutil.PodFitsCurrentNode(d.handle.GetPodsAssignedToNodeFunc(), pod, node) &&
+							nodeutil.PodFitsAnyNode(d.handle.GetPodsAssignedToNodeFunc(), pod, nodes)
+					}),
+				)
+				if err != nil {
+					return &framework.Status{
```

Comment:
To preserve existing behavior, this should log an error message and continue processing next node.

Reply:
I wonder whether getting `err != nil` for the first time when running `podutil.ListPodsOnANode` will only mean there's gonna be `err != nil` the second, the third and the next time `podutil.ListPodsOnANode` gets invoked again. Could be some temporary issue (e.g. connection getting refused) or some hard issue (misconfiguration of the plugin/framework).

Both cases of continuing/not continuing make sense.

Reply:
Do we keep it like this then?